packages/expo-crypto/android/src/main/java/expo/modules/crypto/CryptoModule.kt

```diff
@@ -0,0 +1,49 @@
+package expo.modules.crypto
+
+import android.content.Context
+import android.util.Base64
+
+import expo.modules.core.ExportedModule
+import expo.modules.core.ModuleRegistry
+import expo.modules.core.Promise
+import expo.modules.core.interfaces.ExpoMethod
+
+import java.security.MessageDigest
+import java.security.NoSuchAlgorithmException
+
+class CryptoModule(context: Context?) : ExportedModule(context) {
+  override fun onCreate(moduleRegistry: ModuleRegistry) { }
+
+  override fun getName() = "ExpoCrypto"
+
+  @ExpoMethod
+  fun digestStringAsync(algorithm: String, data: String, options: Map<String, Any?>, promise: Promise) {
+    val encoding = options["encoding"] as String?
+
+    val messageDigest: MessageDigest = try {
+      MessageDigest.getInstance(algorithm).apply { update(data.toByteArray()) }
+    } catch (e: NoSuchAlgorithmException) {
+      promise.reject("ERR_CRYPTO_DIGEST", e)
+      return
+    }
+
+    val digest: ByteArray = messageDigest.digest()
+    when (encoding) {
+      "base64" -> {
+        val output = Base64.encodeToString(digest, Base64.NO_WRAP)
+        promise.resolve(output)
+      }
+      "hex" -> {
+        val output = digest.joinToString(separator = "") {
+          ((it.toInt() and 0xff) + 0x100)
+            .toString(16)
+            .substring(1)
+        }
```

Comment:
L38: Explicit cast from `Byte` to `Int` should be fine. That is almost exactly what Kotlin's `Byte.and` infix operator does: 
```kotlin

/** Performs a bitwise AND operation between the two values. */
@SinceKotlin("1.1")
@kotlin.internal.InlineOnly
public inline infix fun Byte.and(other: Byte): Byte = (this.toInt() and other.toInt()).toByte()
```
(copied from `kotlin.experimental` package)

but we don't want to use it, because it would require the value `0x100` to be casted to `Byte` type, resulting in overflow. 
I checked Java also does widening cast when `byte` is "and'ed" with `int`

Reply:
You're right, it's nothing wrong with converting a `Byte` to `Int` - it's a lossless conversion, because [Byte is 8-bit and Int is 32-bit](https://kotlinlang.org/docs/basic-types.html#integer-types) so it just prepends 24 zeros at the beginning of representation and this doesn't affect the bitwise `and` operation.
After all, the addition with `0x100` (which is already larger than 8-bit) will cast it to `Int` anyway. (_Sidenote: In Kotlin, `Byte + Byte = Int` - it would cause overflows otherwise)_ 

The only thing I'd change here is to use named arguments everywhere, because without copying the code to AS I got lost in what the arguments mean:
```suggestion
      "hex" -> {
        val output = digest.joinToString(separator = "") { byte ->
          ((byte.toInt() and 0xff) + 0x100)
            .toString(radix = 16)
            .substring(startIndex = 1)
        }
```

Reply:
> [...] so it just prepends 24 zeros at the beginning of representation [...] 

Actually this is not strictly true ([link](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-byte/to-int.html)). Rest is correct :smile:  && thanks for good suggestion

Reply:
Yeah 😅 too much working with unsigned integers before narrowed down my thinking to non-negative numbers only, thanks!